Add variants for different build configurations

ig0774 opened the pull request:
This adds a few build variants for some simple configurations and, more importantly, adds a way for custom `.sublime-build` files to override build settings.

Basically, this allows us to implement something not far off what was requested in #255:

<img width="395" alt="screen shot" src="https://cloud.githubusercontent.com/assets/212893/16897588/e856cdc6-4bad-11e6-86d3-38c514a95254.png">

This menu can be brought up by pressing `C-shift-b`. On ST3, after build 3080, this is, of course, simply the default behaviour. However, I've back-ported something similar for earlier versions of ST3 and for ST2. The chosen build variant will persist across builds, but can be reset by choosing the **LaTeX** variant, which is simply the default configuration. (The functionality isn't *exactly* the same, but it's quite close; the main difference is that the selected build is *not* saved across windows or persisted after the application closes).

The other thing this allows is for us to support the request in #743 (see [my comment in that thread](https://github.com/SublimeText/LaTeXTools/issues/743#issuecomment-225304102)). More specifically, the following options can be provided by a `.sublime-build` file:

* `program`
* `builder`
* `command`
* `env`
* `path`

These, if provided, will override the corresponding settings, including the `%!TEX program = ...` magic comment.

We can always add more options if requested, but I thought this was a good starting point.

r-stein commented on latextools_utils/sublime_utils.py:
```diff
@@ -200,102 +202,166 @@ def check_processes(st2_dir=None):
+def parse_json_with_comments(filename):
+    with codecs.open(filename, 'r', 'utf-8', 'ignore') as f:
+        content = f.read()
+
```
For ST3 it might be the best to use `sublime.decode_value(content)`.

ig0774 replied:
Good point.

r-stein commented on makePDF.py:
```diff
@@ -75,7 +84,167 @@ def getOEMCP():
+	if sublime.version() > '3079':
+		def run(self, select=False):
+			select = False if select not in (True, False) else select
+			self.window.run_command('build', {'select': select})
```
I think it would be better to use a context in the keybinding and write an EventListener. There may be packages, which map the `ctrl+b` to something else, e.g. Sublimemacs.

ig0774 replied:
Could you explain a bit more what you have in mind? I'm not sure I follow...

r-stein replied:
For example write an EventListener

``` py
class SublimeVersionListener(sublime_plugin.EventListener):
    def on_query_context(self, view, key, operator, operand, match_all):
        if key != "_lt_st_build_version":
            return
        return sublime.version() <= '3079'
```
And add a context to the keybinding:
``` js
{ "keys": ["ctrl+b"],
  "context": [
    {"key": "selector", "operator": "equal", "operand": "text.tex.latex"}
    {"key": "_lt_st_build_version"}],
  "command": "latextools_build_selector"},
```

This avoids to call the command at all and we don't overwrite keybindings of other packages. I think this is a better solution than calling the other command from this one.

ig0774 commented:
@r-stein: Thanks for the suggestions. They've been incorporated. In terms of the version context mechanism, I thought it was a good idea to make something a bit more generic. Basically, I've implemented the whole set of possible operators for the ST version number. It doesn't allow elegant support of the concept of `<=`, but its close enough.

I also added the ability to override the `script_commands` setting if using the script builder. This seems like a useful feature to cover, e.g., #828.
